app/scripts/lib/rpc-method-middleware/handlers/wallet-requestPermissions.ts

```
@@ -105,65 +98,54 @@ async function requestPermissionsImplementation(
-        requestedPermissions,
-      );
-      grantedPermissions = { ...frozenGrantedPermissions };
-    } catch (error) {
-      return end(error as unknown as Error);
-    }
-  }
+  let grantedPermissions: GrantedPermissions = {};
+
+  const [frozenGrantedPermissions] = await requestPermissionsForOrigin(
+    requestedPermissions,
+  );
 
-  if (caip25Approval) {
-    const grantedCaip25Permissions = grantPermissionsForOrigin(caip25Approval);
-    const caip25Endowment =
-      grantedCaip25Permissions[Caip25EndowmentPermissionName];
+  grantedPermissions = { ...frozenGrantedPermissions };
 
+  if (hasCaip25EquivalentPermissions) {
+    const caip25Endowment = grantedPermissions[Caip25EndowmentPermissionName];
     const caip25CaveatValue = caip25Endowment?.caveats?.find(
       ({ type }) => type === Caip25CaveatType,
     )?.value as Caip25CaveatValue | undefined;
-
     if (!caip25CaveatValue) {
       throw new Error(
         `could not find ${Caip25CaveatType} in granted ${Caip25EndowmentPermissionName} permission.`,
       );
     }
 
+    delete grantedPermissions[Caip25EndowmentPermissionName];
     // We cannot derive correct eth_accounts value directly from the CAIP-25 permission
     // because the accounts will not be in order of lastSelected
     const ethAccounts = getAccounts();
 
-    grantedPermissions[RestrictedMethods.eth_accounts] = {
-      ...caip25Endowment,
-      parentCapability: RestrictedMethods.eth_accounts,
-      caveats: [
-        {
-          type: CaveatTypes.restrictReturnedAccounts,
-          value: ethAccounts,
-        },
-      ],
-    };
+    if (caip25Endowment) {
+      grantedPermissions[RestrictedMethods.eth_accounts] = {
+        ...caip25Endowment,
+        parentCapability: RestrictedMethods.eth_accounts,
+        caveats: [
+          {
+            type: CaveatTypes.restrictReturnedAccounts,
+            value: ethAccounts,
+          },
+        ],
+      };
+    }
```

adding this check here because linter complains about `grantedPermissions[RestrictedMethods.eth_accounts]` possibly not having the required props (id, invoker, date) because of `caip25Endowment` possibly being `undefined`

```tsconst caip25Endowment: ValidPermission<string, Caveat<string, Json>> | undefined````

So I got around this doing this check (similar situation below).

This brought up the question, should we throw an exception in case `caip25Endowment` for some weird reason is `undefined` ?
Similar to what we do in current L124 with 
```ts
    if (!caip25CaveatValue) {
      throw new Error(
        `could not find ${Caip25CaveatType} in granted ${Caip25EndowmentPermissionName} permission.`,
      );
    }
```
@jiexi lmk your thoughts here

---

For what it's worth, I think we should throw an error in that case, since it's a possibility.

---

done in https://github.com/MetaMask/metamask-extension/pull/30042/commits/6adaff6a6fdf63f11a46da119d7d1df1a5eafb9b

---

doing this removes the need for checking if `caip25Endowment` is defined, so removed check in https://github.com/MetaMask/metamask-extension/pull/30042/commits/f877d0bcff08df72ab77a59021ffeba336b87a2a